File: src/main/java/work/alsace/alsacecore/Util/WarpDataLoader.java
package work.alsace.alsacecore.Util;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.configuration.file.YamlConfiguration;
import work.alsace.alsacecore.AlsaceCore;

import java.io.File;
import java.io.IOException;
import java.util.*;

public class WarpDataLoader {
    private final Map<String, Location> warps = Collections.synchronizedMap(new HashMap<>());
    private final Map<String, String> aliases = Collections.synchronizedMap(new HashMap<>());
    private final File warpsFile;
    private YamlConfiguration warpConfig;

    public WarpDataLoader(AlsaceCore plugin) {
        this.warpsFile = new File(plugin.getDataFolder() + File.separator + "warps.yml");
        this.loadWarps();
    }

    private synchronized void loadWarps() {
        if (!warpsFile.exists()) {
            try {
                warpsFile.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        warpConfig = YamlConfiguration.loadConfiguration(warpsFile);

        if (warpConfig.getConfigurationSection("Warps") != null) {
            for (String i : warpConfig.getConfigurationSection("Warps").getKeys(false)) {
                warps.put(i, new Location(
                        Bukkit.getWorld(Objects.requireNonNull(warpConfig.getString("Warps." + i + ".World"))),
                        warpConfig.getDouble("Warps." + i + ".X"),
                        warpConfig.getDouble("Warps." + i + ".Y"),
                        warpConfig.getDouble("Warps." + i + ".Z"),
                        warpConfig.getLong("Warps." + i + ".Yaw"),
                        warpConfig.getLong("Warps." + i + ".Pitch")));

                // 加载别名
                String alias = warpConfig.getString("Aliases." + i);
                if (alias != null) {
                    aliases.put(alias, i);
                }
            }
        }
    }

    public void addWarp(String name, String alias, Location location) {
        synchronized (warps) {
            warps.put(name, location);
            warpConfig.set("Warps." + name + ".World", Objects.requireNonNull(location.getWorld()).getName());
            warpConfig.set("Warps." + name + ".X", location.getX());
            warpConfig.set("Warps." + name + ".Y", location.getY());
            warpConfig.set("Warps." + name + ".Z", location.getZ());
            warpConfig.set("Warps." + name + ".Yaw", location.getYaw());
            warpConfig.set("Warps." + name + ".Pitch", location.getPitch());

            if (alias != null && !alias.isEmpty()) {
                warpConfig.set("Aliases." + alias, name);
                aliases.put(alias, name);
            }

            try {
                warpConfig.save(warpsFile);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public void delWarp(String name) {
        synchronized (warps) {
            warps.remove(name);
            warpConfig.set("Warps." + name, null);

            // 删除别名
            String alias = getWarpAlias(name);
            if (alias != null) {
                warpConfig.set("Aliases." + alias, null);
                aliases.remove(alias);
            }

            try {
                warpConfig.save(warpsFile);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public Location getWarp(String nameOrAlias) {
        synchronized (warps) {
            Location location = warps.get(nameOrAlias);
            if (location == null) {
                String realName = aliases.get(nameOrAlias);
                if (realName != null) {
                    location = warps.get(realName);
                }
            }
            return location;
        }
    }

    public Set<String> getWarps() {
        return new HashSet<>(warps.keySet());
    }

    public World getWarpWorld(String name) {
        Location location = getWarp(name);
        return (location != null) ? location.getWorld() : null;
    }

    public String getWarpAlias(String name) {
        return aliases.get(name);
    }

    public String getRealWarpName(String alias) {
        return aliases.getOrDefault(alias, alias);
    }
}

File: src/main/java/work/alsace/buildercore/service/WarpDataLoader.kt
package work.alsace.buildercore.service

import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.World
import org.bukkit.configuration.file.YamlConfiguration
import work.alsace.buildercore.BuilderCore
import java.io.File
import java.io.IOException
import java.util.*

class WarpDataLoader(plugin: BuilderCore) {
    private val warps = Collections.synchronizedMap(HashMap<String, Location>())
    private val aliases = Collections.synchronizedMap(HashMap<String, String>())
    private val warpsFile: File
    private var warpConfig: YamlConfiguration? = null

    init {
        warpsFile = File(plugin.dataFolder.toString() + File.separator + "warps.yml")
        loadWarps()
    }

    @Synchronized
    private fun loadWarps() {
        if (!warpsFile.exists()) {
            try {
                warpsFile.createNewFile()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        warpConfig = YamlConfiguration.loadConfiguration(warpsFile)
        if (warpConfig!!.getConfigurationSection("Warps") != null) {
            for (i in Objects.requireNonNull(warpConfig!!.getConfigurationSection("Warps"))?.getKeys(false)!!) {
                warps[i] = Location(
                    Objects.requireNonNull(warpConfig!!.getString("Warps.$i.World"))?.let { Bukkit.getWorld(it) },
                    warpConfig!!.getDouble("Warps.$i.X"),
                    warpConfig!!.getDouble("Warps.$i.Y"),
                    warpConfig!!.getDouble("Warps.$i.Z"),
                    warpConfig!!.getLong("Warps.$i.Yaw").toFloat(),
                    warpConfig!!.getLong("Warps.$i.Pitch").toFloat()
                )

                // 加载别名
                val alias = warpConfig!!.getString("Aliases.$i")
                if (alias != null) {
                    aliases[alias] = i
                }
            }
        }
    }

    fun addWarp(name: String, alias: String?, location: Location) {
        synchronized(warps) {
            warps[name] = location
            warpConfig!!["Warps.$name.World"] = Objects.requireNonNull(location.world)?.name
            warpConfig!!["Warps.$name.X"] = location.x
            warpConfig!!["Warps.$name.Y"] = location.y
            warpConfig!!["Warps.$name.Z"] = location.z
            warpConfig!!["Warps.$name.Yaw"] = location.yaw
            warpConfig!!["Warps.$name.Pitch"] = location.pitch
            if (!alias.isNullOrEmpty()) {
                warpConfig!!["Aliases.$alias"] = name
                aliases[alias] = name
            }
            try {
                warpConfig!!.save(warpsFile)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    fun delWarp(name: String) {
        synchronized(warps) {
            warps.remove(name)
            warpConfig!!["Warps.$name"] = null

            // 删除别名
            val alias = getWarpAlias(name)
            if (alias != null) {
                warpConfig!!["Aliases.$alias"] = null
                aliases.remove(alias)
            }
            try {
                warpConfig!!.save(warpsFile)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    fun getWarp(nameOrAlias: String): Location? {
        synchronized(warps) {
            var location = warps[nameOrAlias]
            if (location == null) {
                val realName = aliases[nameOrAlias]
                if (realName != null) {
                    location = warps[realName]
                }
            }
            return location
        }
    }

    fun getWarps(): Set<String> {
        return HashSet(warps.keys)
    }

    fun getWarpWorld(name: String): World? {
        val location = getWarp(name)
        return location?.world
    }

    fun getWarpAlias(name: String): String? {
        return aliases[name]
    }

    fun getRealWarpName(alias: String): String {
        return aliases.getOrDefault(alias, alias)
    }
}
